[ParseableInterface] Serialize module into buffer and load from it

harlanhaskins opened the pull request:
In addition to being wasteful, this is a correctness issue -- the
compiler should only ever have one view of this file, and it should not
read a potentially different file after validating dependencies.

rdar://48654608

harlanhaskins committed:
[ParseableInterface] Only open module buffers once while loading

In addition to being wasteful, this is a correctness issue -- the
compiler should only ever have one view of this file, and it should not
read a potentially different file after validating dependencies.

rdar://48654608

Azoy commented:
Really amazing comments in this patch!

harlanhaskins commented:
@Azoy Thanks!

harlanhaskins commented:
@swift-ci please smoke test

jrose-apple commented on lib/Frontend/ParseableInterfaceModuleLoader.cpp:
```diff
@@ -173,30 +173,40 @@ class DiscoveredModule {
-  static DiscoveredModule normal(StringRef path) {
-    return { path, Kind::Normal };
+  static DiscoveredModule normal(StringRef path,
+    std::unique_ptr<llvm::MemoryBuffer> moduleBuffer) {
```
Nitpick: double-indent wrapped lines (per clang-format).

jrose-apple commented on lib/Serialization/Serialization.cpp:
```diff
@@ -4809,6 +4810,48 @@ void Serializer::writeToStream(raw_ostream &os, ModuleOrSourceFile DC,
   S.writeToStream(os);
 }
 
+void swift::serializeToBuffers(
```
Can you implement `serialize` in terms of this, just so we don't have two implementations?

harlanhaskins replied:
Are we cool with double-buffering the serialization?

jrose-apple replied:
Oh, right. It won't matter once we get a chance to do the mmap thing, but it does now.

jrose-apple commented on lib/Serialization/Serialization.cpp:
```diff
@@ -48,6 +48,7 @@
 #include "llvm/Support/EndianStream.h"
 #include "llvm/Support/FileSystem.h"
 #include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/SmallVectorMemoryBuffer.h"
```
Nitpick: alphabet

jrose-apple commented on lib/Frontend/ParseableInterfaceModuleLoader.cpp:
```diff
@@ -559,7 +570,14 @@ class swift::ParseableInterfaceBuilder {
+          // We don't want to serialize module docs in the cache -- they
+          // will be serialized beside the interface file.
+          serializeToBuffers(Mod, SerializationOpts, ModuleBuffer,
+                             /*ModuleDocBuffer*/nullptr, SILMod.get());
```
Seems like we could always pass down `ModuleBuffer` and `nullptr` and get the same effect?

harlanhaskins replied:
I thought about that, but I preferred explicitly encoding the difference in behavior. But if you think they should be combined, I'll combine them.

jrose-apple replied:
I feel like the less a function changes its behavior based on inputs, the better.

jrose-apple commented on include/swift/Subsystems.h:
```diff
@@ -264,6 +264,17 @@ namespace swift {
 
+  /// Serializes a module or single source file to the given output file and
+  /// returns back the file's contents as a memory buffer.
+  /// Use this if you intend to immediately load the serialized module, as that
```
Docs nitpick: this would be a good place for a newline to separate the "short" and "full" docs.

jrose-apple reviewed (commented):
I like the safety of `std::unique_ptr` a lot here. Ownership FTW!

jrose-apple commented on lib/Serialization/Serialization.cpp:
```diff
@@ -4839,7 +4839,7 @@ void swift::serializeToBuffers(
     llvm::SmallString<1024> buf;
     llvm::raw_svector_ostream stream(buf);
-    Serializer::writeToStream(stream, DC, M, options);
+    writeDocToStream(stream, DC, options.GroupInfoPath);
```
Heh, good catch.

benlangmuir reviewed (approved):
LGTM, thanks Harlan!

harlanhaskins committed:
[ParseableInterface] Add test for writing modules to a buffer

harlanhaskins commented:
@swift-ci please test

harlanhaskins committed:
[ClangImporter] Allow importing without a resource dir on non-Darwin

Previously, we would assert that there's a runtime resource dir
available, but then accept the possibility of no glibc.modulemap. We
should just do the same thing for 'no resource dir' as 'no module map'.

harlanhaskins commented:
@swift-ci please smoke test Linux

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -571,28 +600,10 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
-        (Twine("-fmodule-map-file=") + GlibcModuleMapPath).str());
+    if (auto path = getGlibcModuleMapPath(searchPathOpts.RuntimeResourcePath,
+                                          triple, GlibcModuleMapPath)) {
+      invocationArgStrs.push_back(*path);
```
You lost the `-fmodule-map-file=`.

harlanhaskins commented:
@swift-ci please smoke test Linux

harlanhaskins commented:
@swift-ci please smoke test macOS

harlanhaskins commented:
Seems like the Linux bot had issues. Let's try again.

@swift-ci please smoke test Linux

harlanhaskins commented:
@swift-ci please test and merge

harlanhaskins commented:
@swift-ci please test
